recipes-core/rauc/files/nvbootctrl-rauc

```diff
@@ -0,0 +1,110 @@
+#!/bin/sh
+#
+# nvbootctrl-rauc: rauc script to get/set the bootloader status
+#
+# SPDX-License-Identifier: MIT
+
+usage() {
+    cat >&2 <<EOF
+Usage: $0 <ACTION> <BOOTNAME> <STATUS>
+Options:
+    -h,                             Print this usage message
+    ACTION
+Actions:
+    get-current                     Returns the current booted bootname
+    get-primary                     Returns the primary bootname
+    get-state <BOOTNAME>            Returns the BOOTNAME status ("good", "bad")
+    set-primary <BOOTNAME>          Set the primary slot to BOOTNAME (needs reboot)
+    set-state <BOOTNAME> <STATUS>   Apply the STATUS ("good", "bad") to BOOTNAME
+EOF
+exit 1; }
+
+SLOT_A="APP"
+SLOT_B="APP_b"
+
+function get_status(){
+    { nvbootctrl dump-slots-info | grep "slot: ${1}" | grep -q "normal"; } && \
+    { nvbootctrl -trootfs dump-slots-info | grep "slot: ${1}" | grep -q "normal"; } && \
+    echo 0 || echo 1
+}
+
+function get_slot(){
+    case "${1}" in
+        *_b)
+            echo 1;
+            ;;
+        *)
+            echo 0;
+            ;;
+    esac
+}
+
+function get_slot_AB(){
+    case "${1}" in
+        *_b)
+            echo B;
+            ;;
+        *)
+            echo A;
+            ;;
+    esac
+}
+
+function get_slot_err_code(){
+    case "${1}" in
+        good)
+            echo "00";
+            ;;
+        bad)
+            echo "FF";
+            ;;
+    esac
+}
+
+while getopts ":h" o; do
+    case "${o}" in
+        *)
+            usage
+            ;;
+    esac
+done
+shift $((OPTIND-1))
+
+if [ -z "${1}" ] ; then
+    usage
+fi
+
+case "${1}" in
+    get-current|get-primary)
+        [ $(nvbootctrl get-current-slot) == "0" ] && echo "${SLOT_A}" || echo "${SLOT_B}"
+        exit 0
+        ;;
+    get-state)
+        if [ -z "${2}" ] ; then
+            usage
+        fi
+        SLOT="$(get_slot ${2})"
+        [ "$(get_status ${SLOT})" == "0" ] && echo "good" || echo "bad"
+        exit 0
+        ;;
+    set-primary)
+        if [ -z "${2}" ] ; then
+            usage
+        fi
+        SLOT="$(get_slot ${2})"
+        nvbootctrl set-active-boot-slot "${SLOT}"; exit $?
+        ;;
+    set-state)
+        if [ -z "${2}" ] || [ -z "${3}" ] ; then
+            usage
+        fi
+        SLOT="$(get_slot_AB ${2})"
+        ERRCODE="$(get_slot_err_code ${3})"
+        printf "\x07\x00\x00\x00\x${ERRCODE}\x00\x00\x00" > \
+            "/opt/nvidia/esp/EFI/NVDA/Variables/RootfsStatusSlot${SLOT}-781e084c-a330-417c-b678-38e696380cb9"
```

Comment:
This method of writing EFI variables does not work on the AGX Xavier.  You might instead want to use the `set_efi_var` function in the `uefi_common.func` file installed by the `setup-nv-boot-control` recipe.

Reply:
@madisongh @joekale-pp:  Hi, I'am currently trying to integrate RAUC and I came across your recommendation to use the set_efi_var function provided in the uefi_common.func file installed by the setup-nv-boot-control recipe.
Could you please provide clarification on:
- Why the direct method of writing EFI variables is limited to the AGX Xavier platform?
- How does the set_efi_var function address this limitation?

Reply:
Just take a look at [the code](https://github.com/OE4T/meta-tegra/blob/master/recipes-bsp/tools/setup-nv-boot-control/uefi_common.func.in) to see how it works.

As for why, it's apparently due to the AGX Xavier platform using eMMC storage only.  All the others use SPI flash for the boot firmware, and NVIDIA's UEFI implementation for that configuration supports the normal EFI variable interfaces.